File: ml6-d-svn/trunk/mdl/HolmesPrasad/nonMDL/Matrix.java
package nonMDL;

/*========== Matrix.java ==========
  Matrix will hold a 2-d array of doubles and have a default size of 4x4.
  Handles basic matrix maintenence and math.
  Creates transformation matricies for tralation, scale and rotate
=========================*/

import java.io.*;
import java.util.*;

public class Matrix {

    public static final int DEFAULT_SIZE = 4;
    protected double[][] m;

    /*===========Constructors================
      Default constructor creates a 4x4 matrix
      Second constructor creates a 4xN matrix
    */
    public Matrix() {
	m = new double[DEFAULT_SIZE][DEFAULT_SIZE];
    }
    public Matrix(int c) {
	m = new double[DEFAULT_SIZE][c];
    }

    /*===========grow================
      Increase the number of columns in a matrix by 10
      You can change the growth factor as you see fit
      Puts old matrix starting in the upper left corner
    */
    public void grow() {

	double[][] n = new double[m.length][m[0].length + 10];
	for (int r=0; r<m.length; r++)
	    for (int c=0; c<m[r].length; c++)
		n[r][c] = m[r][c];
	
	m = n;
    }

    /*======== public void clear() ==========
      Inputs:  
      Returns: 
      Sets every entry in the matrix to 0
      ====================*/
    public void clear() {

	for (int i=0; i<m.length; i++) 
	    for (int j=0; j<m[i].length; j++) 
		m[i][j] = 0;
    }

    /*===========ident================
      Turns this matrix into the indentity matrix
      You may assume the calling Matrix is square
    */
    public void ident() {
	for (int r = 0; r < m.length; r++)
	    for (int c = 0; c < m[0].length; c++)
		m[r][c] = r == c ? 1 : 0;
    }

    /*===========scalarMult================
      Inputs:  double s
      
      multiply each element of the calling matrix by s
    */
    public void scalarMult( int s ) {
	for (int r = 0; r < m.length; r++)
	    for (int c = 0; c < m[0].length; c++)
		m[r][c] *= s;
    }		

    /*===========matrixMult================
      Multply matrix n by the calling matrix, modify
      the calling matrix to store the result.
      
      eg.
      In the call a.matrixMult(n), n will remain the same
      and a will now be the product of n * a
    */
    public void matrixMult( Matrix n ) {
	if (n.m[0].length != this.m.length){
	    throw new UnsupportedOperationException("the number of columns in left matrix doesn't match the number of rows in the right matrix");
	}
	double[][] r = new double[n.m.length][this.m[0].length];
	for (int rr = 0; rr < r.length; rr++)
	    for (int rc = 0; rc < r[0].length; rc++){
		r[rr][rc] = 0;
		for (int i = 0; i < this.m.length; i++)
		    r[rr][rc] += n.m[rr][i] * this.m[i][rc];
	    }
	this.m = r;
    }
   
    /*===========copy================
      Create and return new matrix that is a duplicate 
      of the calling matrix
    */
    public Matrix copy() {
	Matrix returner = new Matrix(m[0].length);
	for (int r = 0; r < m.length; r++)
	    for (int c = 0; c < m[0].length; c++)
		returner.m[r][c] = this.m[r][c];
	return returner;
    }

    /*===========toString================
      Create a readable String representation of the 
      calling matrix.
    */
    public String toString() {
	StringBuilder sb = new StringBuilder();
	Formatter f = new Formatter (sb, Locale.US);
	for (int r = 0; r < m.length; r++){
	    for (int c = 0; c < m[0].length; c++){
		f.format("%1$12.4g ", m[r][c]);
	    }
	    f.format("%n");
	}
	return sb.toString();
    }
    /*===========MakeTranslate================
      Turns the calling matrix into the appropriate
      translation matrix using x, y, and z as the translation
      offsets.
     */
    public void MakeTranslate(double x, double y, double z) {
	ident();
	m[0][3] = x;
	m[1][3] = y;
	m[2][3] = z;
    }
    
    /*===========MakeScale================
      Turns the calling matrix into the appropriate scale
      matrix using x, y and z as the scale factors.
     */
    public void MakeScale(double x, double y, double z) {
	// ident();  // unnecessary because scale factors go on the diagonal, where the 1s would go
	m[0][0] = x;
	m[1][1] = y;
	m[2][2] = z;
    }

    /*=========== MakeRotX ================
      Turns the calling matrix into the appropriate rotation
      matrix using theta as the angle of rotation and X
      as the axis of rotation.  RADIANS
    */
    public void MakeRotX(double theta){
	ident();
	m[1][1] = Math.cos(theta);
	m[1][2] = -Math.sin(theta);
	m[2][1] = Math.sin(theta);
	m[2][2] = Math.cos(theta);
    }

    /*=========== MakeRotY ================
      Turns the calling matrix into the appropriate rotation
      matrix using theta as the angle of rotation and Y
      as the axis of rotation.  RADIANS
    */
    public void MakeRotY(double theta) {
	ident();
	m[0][0] = Math.cos(theta);
	m[0][2] = Math.sin(theta);
	m[2][0] = -Math.sin(theta);
	m[2][2] = Math.cos(theta);
    }

    /*=========== MakeRotZ ================
      Turns the calling matrix into the appropriate rotation
      matrix using theta as the angle of rotation and axis
      as the axis of rotation.  RADIANS
    */
    public void MakeRotZ(double theta) {
	ident();
	m[0][0] = Math.cos(theta);
	m[0][1] = -Math.sin(theta);
	m[1][0] = Math.sin(theta);
	m[1][1] = Math.cos(theta);
    }
}

File: ml6-d-svn/trunk/transformations/AndyChen/matrix.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#include "matrix.h"

/*-------------- struct matrix *new_matrix() --------------
Inputs:  int rows
         int cols 
Returns: 

Once allocated, access the matrix as follows:
m->m[r][c]=something;
if (m->lastcol)... 
*/
struct matrix *new_matrix(int rows, int cols) {
  double **tmp;
  int i;
  struct matrix *m;

  tmp = (double **)malloc(rows * sizeof(double *));
  for (i=0;i<rows;i++) {
      tmp[i]=(double *)malloc(cols * sizeof(double));
    }

  m=(struct matrix *)malloc(sizeof(struct matrix));
  m->m=tmp;
  m->rows = rows;
  m->cols = cols;
  m->lastcol = 0;

  return m;
}


/*-------------- void free_matrix() --------------
Inputs:  struct matrix *m 
Returns: 

1. free individual rows
2. free array holding row pointers
3. free actual matrix
*/
void free_matrix(struct matrix *m) {

  int i;
  for (i=0;i<m->rows;i++) {
      free(m->m[i]);
    }
  free(m->m);
  free(m);
}


/*======== void grow_matrix() ==========
Inputs:  struct matrix *m
         int newcols 
Returns: 

Reallocates the memory for m->m such that it now has
newcols number of collumns
====================*/
void grow_matrix(struct matrix *m, int newcols) {
  int i;
  for (i=0;i<m->rows;i++) {
      m->m[i] = realloc(m->m[i],newcols*sizeof(double));
    }
  m->cols = newcols;
}


/*-------------- void print_matrix() --------------
Inputs:  struct matrix *m 
Returns: 

print the matrix
*/
void print_matrix(struct matrix *m) {
  int i,j;
  for (i = 0; i< m->rows; i++)
    {for (j=0; j<m->cols; j++)
        {printf("%10lf",m->m[i][j]);
        }
      printf("\n");
    }
  printf("\n");
}

/*-------------- void ident() --------------
Inputs:  struct matrix *m <-- assumes m is a square matrix
Returns: 

turns m in to an identity matrix
*/
void ident(struct matrix *m) {
  int i,j;
  for (i = 0; i< m->rows; i++)
    {for (j=0; j<m->cols; j++)
        {if (i==j)
            m->m[i][j] = 1;
            
         else
            m->m[i][j] = 0;
            
        }
      
    }

}


/*-------------- void scalar_mult() --------------
Inputs:  double x
         struct matrix *m 
Returns: 

multiply each element of m by x
*/
void scalar_mult(double x, struct matrix *m) {
  int i,j;
  for (i = 0; i< m->rows; i++)
    {for (j=0; j<m->cols; j++)
        {m->m[i][j] = m->m[i][j]*x;
        }
     
    }
}


/*-------------- void matrix_mult() --------------
Inputs:  struct matrix *a
         struct matrix *b 
Returns: 

a*b -> b
*/
void matrix_mult(struct matrix *a, struct matrix *b) {
  struct matrix *temp;
        temp = new_matrix(b->rows, b->cols);
        
        int i, j, k;
        double sum;
        i = j = k = sum = 0;
        while (i < b->rows){
                j = 0;
                while (j < b-> cols){
                        sum = k = 0;
                        while (k < b->rows && k < b->cols){
                                sum += a->m[i][k] * b->m[k][j];
                                k++;
                        }
                        temp->m[i][j] = sum;
                        j++;
                }
                i++;
        }
       
        
        copy_matrix(temp, b);
        free_matrix(temp);

 
}



/*-------------- void copy_matrix() --------------
Inputs:  struct matrix *a
         struct matrix *b 
Returns: 

copy matrix a to matrix b
*/
void copy_matrix(struct matrix *a, struct matrix *b) {
  int i, j;
  i = j = 0;
  while (i < a->rows){
    j = 0;
    while (j < a->cols){
      b->m[i][j] = a->m[i][j];
      j++;
    }
    i++;
  }

}


/*======== struct matrix * make_translate() ==========
Inputs:  int x
         int y
         int z 
Returns: The translation matrix created using x, y and z 
as the translation offsets.
====================*/
struct matrix * make_translate(double x, double y, double z) {
struct matrix *a;
a = new_matrix(4,4);
ident(a);
a->m[0][3]=x;
a->m[1][3]=y;
a->m[2][3]=z;

return a;
}

/*======== struct matrix * make_scale() ==========
Inputs:  int x
         int y
         int z 
Returns: The translation matrix creates using x, y and z
as the scale factors
====================*/
struct matrix * make_scale(double x, double y, double z) {
struct matrix *a;
a = new_matrix(4,4);
ident(a);
a->m[0][0]=x;
a->m[1][1]=y;
a->m[2][2]=z;
return a;
}

/*======== struct matrix * make_rotX() ==========
Inputs:  double theta

Returns: The rotation matrix created using theta as the 
angle of rotation and X as the axis of rotation.
====================*/
struct matrix * make_rotX(double theta) {
  struct matrix *a;
  double radians = (theta / 180)*M_PI;
  a = new_matrix(4,4);
  ident(a);
  a->m[1][1]=cos(radians);
  a->m[1][2]=-1 * sin(radians);
  a->m[2][1]=sin(radians);
  a->m[2][2]=cos(radians);
  return a;
}

/*======== struct matrix * make_rotY() ==========
Inputs:  double theta
         char c 
Returns: The rotation matrix created using theta as the 
angle of rotation and Y as the axis of rotation.
====================*/
struct matrix * make_rotY(double theta) {
  struct matrix *a;
  double radians = (theta / 180)*M_PI;
  a = new_matrix(4,4);
  ident(a);
  a->m[0][0]=cos(radians);
  a->m[0][2]=sin(radians);
  a->m[2][0]=-1 * sin(radians);
  a->m[2][2]=cos(radians);
  return a;
}

/*======== struct matrix * make_rotZ() ==========
Inputs:  double theta
         char c 
Returns: The rotation matrix created using theta as the 
angle of rotation and Z as the axis of rotation.
====================*/
struct matrix * make_rotZ(double theta) {
  struct matrix *a;
  double radians = (theta / 180)*M_PI;
  a = new_matrix(4,4);
  ident(a);
  a->m[0][0]=cos(radians);
  a->m[0][1]=-1 * sin(radians);
  a->m[1][0]=sin(radians);
  a->m[1][1]=cos(radians);
  return a;
}
